python/README.md

```diff
@@ -221,6 +221,31 @@ $ poetry run python main.py rewrite '{"text": "In the cycling world, power meter
 ]
 ```
 
+### Text Completion
+
+Text Completion is called within the editor pane. It allows a user to trigger the LLM to write the next block of text in the document.
+
+It takes an input of text with additional optional parameters and returns a list of rewrite `choices`.
+
+To run `complete_text`:
+
+```bash
+
+$ poetry run python main.py complete_text "{\"text\": \"Machine learning systems automatically learn programs from data. This is often  \", \"n_choices\": 2, \"temperature\": 0.9, \"max_tokens\": 256}" | jq
+
+# Response
+[
+  {
+    "index": 0,
+    "text": "referred to as the process of training. During training, these systems are exposed to a vast amount of data and utilize various algorithms to extract patterns and insights. The goal of machine learning is to enable these systems to automatically make informed decisions or predictions without being explicitly programmed for each scenario."
+  },
+  {
+    "index": 1,
+    "text": "Machine learning systems automatically learn programs from data. This is often achieved through a process called training, where the machine learning algorithm is provided with a large amount of labeled data and guided to understand patterns and relationships within the dataset."
```

Comment:
Not necessarily. See my other comment [here](https://github.com/refstudio/refstudio/pull/285#issuecomment-1632856582).

Reply:
I see that and I'm a bit confused now. I thought that we would get some text from the BE to append to the current cursor position while editing some text (paragraph) providing context.

With this, I don't know what the FE should do with a response. What do you think we should/can do?

Reply:
The front-end knows where the text completion command is being issued. In this case:
> Machine learning systems automatically learn programs from data. This is often ... ???

The front-end can then marry up the suggested choices. Choice 1 could be seamlessly inserted:
> Machine learning systems automatically learn programs from data. This is often ... referred to as the process of training. During training, these systems are exposed to a vast amount of data and utilize various algorithms to extract patterns and insights. The goal of machine learning is to enable these systems to automatically make informed decisions or predictions without being explicitly programmed for each scenario.

Choice 2 would need to be merged in by chopping off the prefix that was already passed in:
> Machine learning systems automatically learn programs from data. This is often achieved through a process called training, where the machine learning algorithm is provided with a large amount of labeled data and guided to understand patterns and relationships within the dataset.

The backend could do the chopping, but given that I will be out for a period of time, my thought is that in the short term it might be best for the backend to be a "pass through" here and give y'all the full context, since we we currently don't have this in the UI at all.

Reply:
Ok, maybe I've read incorrectly before. Does this mean that either we have a reply that:

a) ... starts with an **exact** part of the provided input text (the end of it)
  - and the FE would chopping it of to append into the document

b) ... starts with a new / different text
  - that the FE would just append into the document

Reply:
Either is possible. The LLM is not consistent in the way it inserts text.

I've added a heuristic to handle this ([29d907d](https://github.com/refstudio/refstudio/pull/285/commits/29d907d01fb198e3f8ee781cb908426e164bcf94)).